### PR TITLE
jwt_authn: bypass CORS preflight request

### PR DESCRIPTION
**Pull request by yangminzhu**
Signed-off-by: Yangmin Zhu <ymzhu@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: Bypass the CORS preflight request in the JWT filter
Risk Level: Low
Testing: Added unit test and integration test
Docs Changes: n/a
Release Notes: Added `jwt_authn: added to bypass the CORS preflight request.`
Optional Fixes: https://github.com/istio/istio/issues/16171
[Optional Deprecated:]

### CODE REVIEW

**Comment by yangminzhu** [comment]
cc @qiwzhang @lizan

**Comment by yangminzhu** [comment]
@lizan @qiwzhang @diemtvu 

I have a few questions regarding the CORS preflight support in the jwt_authn filter and it might be faster to just ask in the PR.

1) Can we add the support to allow the jwt_authn filter to bypass the CORS preflight request? This unblocks the use case in https://github.com/istio/istio/issues/16171 and might be useful for other Envoy customers as well. 

There might be a alternative solution that doesn't change the jwt_authn filter by adding a special `RequirementRule` that sets `allow_missing_or_failed` to true for CORS preflight request.

2) I prefer to add this as the bulit-in feature in the jwt_authn filter as it's kind of a standard to bypass the CORS preflight request. This PR adds the support without any configuration changes. If you prefer to add a new option to control this, we can add something like `bool bypass_cors_preflight` to `JwtAuthentication`.

3) Let me know if you have any other suggestions, thank you!

**Comment by diemtvu** [comment]
For the Istio issue, it should be solved with authz (rbac), or the custom
authn filter today.  Having jwt filter always passed, preflight request not
going to help if other filters still reject it.

I let @lizan and @qiwzhang address the need for Envoy. Personally, I think
we should simplify this JWT filter and move more to rbac for access
control.

File: source/extensions/filters/http/jwt_authn/filter.cc [hunk 1]
```diff
@@ -29,11 +30,29 @@ void Filter::onDestroy() {
...
+         !headers.AccessControlRequestMethod()->value().empty();
+}
+
 Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool) {
   ENVOY_LOG(debug, "Called Filter : {}", __func__);
 
   state_ = Calling;
   stopped_ = false;
+
+  if (isCorsPreflightRequest(headers)) {
```
Comment (qiwzhang) [inline source/extensions/filters/http/jwt_authn/filter.cc]: This can be easily done by adding a Requiretment Rule.   I prefer not to hard code this.

**Comment by qiwzhang** [comment]
I prefer 2) with a flag in config "bypass_cors_preflight" to enable it.

**Comment by yangminzhu** [comment]
@diemtvu Actually this PR doesn't fix the bug in Istio, I have another PR that changes the istio_authn filter.

@lizan Let me know what do you think.

**Comment by lizan** [comment]
Can you fix the doc build?

**Comment by yangminzhu** [comment]
@lizan Fixed the doc, thanks for the review.

File: source/extensions/filters/http/jwt_authn/filter.cc [hunk 2]
```diff
@@ -29,11 +30,29 @@ void Filter::onDestroy() {
   }
 }
 
+bool isCORSPreflightRequest(const Http::HeaderMap& headers) {
```
Comment (lizan) [inline source/extensions/filters/http/jwt_authn/filter.cc]: nit:
```suggestion
bool isCorsPreflightRequest(const Http::HeaderMap& headers) {
```

Comment (lizan) [inline source/extensions/filters/http/jwt_authn/filter.cc]: put this in unnamed namespace?

File: source/extensions/filters/http/jwt_authn/filter.cc [hunk 3]
```diff
@@ -29,11 +30,29 @@ void Filter::onDestroy() {
...
+         !headers.AccessControlRequestMethod()->value().empty();
+}
+
 Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool) {
   ENVOY_LOG(debug, "Called Filter : {}", __func__);
 
   state_ = Calling;
   stopped_ = false;
+
+  if (config_->bypassCORSPreflightRequest() && isCORSPreflightRequest(headers)) {
```
Comment (lizan) [inline source/extensions/filters/http/jwt_authn/filter.cc]:
```suggestion
  if (config_->bypassCorsPreflightRequest() && isCorsPreflightRequest(headers)) {
```

**Comment by htuch** [comment]
@envoyproxy/api-shepherds API LGTM

**Pull request merged**